pkg/storage/mvcc.go

```diff
@@ -3777,10 +3780,11 @@ func MVCCPredicateDeleteRange(
 			if runSize < rangeTombstoneThreshold {
 				// Only buffer keys if there's a possibility of issuing point tombstones.
 				//
-				// To avoid unecessary memory allocation, overwrite the previous key at
-				// buffer's current position. No data corruption occurs because the
-				// buffer is flushed up to runSize.
-				buf[runSize] = append(buf[runSize][:0], runEnd...)
+				// TODO(ssd): It would be nice to avoid unnecessary allocations
+				// here. We copy the end key because when mvccInternalPut publishes
+				// the logical operation related to the delete, it currenty assumes
+				// that the key will not be subsequently modified.
+				buf = append(buf, runEnd.Clone())
```

Comment:
@msbutler I think you are familiar with this code.  Let me know if you have an opinion here on a good allocation strategy.   Maybe instead of this, perhaps we keep the overwritting, but then allocate a whole new buf anytime we flush point deletions. That way we don't pay allocation costs in the case that we were able to generate range deletes. Do we have a benchmark for this function anywhere?

Reply:
I'm unsure which strategy wins. If we see a pattern of many switches between point key runs and range key runs, then I think your current PR wins, while a pattern of fewer switches would prefer your alternative approach. I didn't write a benchmark function anywhere, and it seems would should add one that looks like https://github.com/cockroachdb/cockroach/blob/2b18f4ffc260dca938ba0265217c4483ec6064c7/pkg/storage/bench_pebble_test.go#L432